flocker/provision/_libcloud.py

```diff
@@ -222,3 +229,150 @@ def create_node(self, name, distribution,
             node=node, address=public_address,
             private_address=private_address,
             distribution=distribution)
+
+    def _create_with_retry(self, **kwargs):
+        """
+        Create a compute instance.  Clean up failed attempts and retry if
+        necessary.
+        """
+        return poll_until(lambda: self._get_node(**kwargs), iter([15] * 10))
+
+    def _get_node(self, name, **kwargs):
+        """
+        Create a compute instance.
+
+        :return: If the node is created successfully, a two-tuple of the
+            libcloud node object and the instance's public IP address.  If an
+            error occurs creating the node, ``None``.
+        """
+        try:
+            node = self._driver.create_node(name=name, **kwargs)
+        except:
+            # We don't know if we just created a node or not.  Look for it and
+            # destroy it if we did.  Hopefully we don't encounter too many more
+            # errors on the way.
+            self._cleanup_node_named(name)
+            raise
+
+        # libcloud has wait_until_running but it doesn't understand the error
+        # state that Rackspace instances often go in to.  It will keep waiting
+        # for the instance to get to the running state even after it has gone
+        # to the error state.  Eventually it will time out, but only after
+        # waiting for much longer than necessary.
+        #
+        # So do a loop ourselves.
+
+        # Here are the states that indicate we've either succeeded or failed.
+        # Only RUNNING indicates success.  The rest are failure states.  Once
+        # the node is in one of those states, it makes no sense to keep
+        # waiting.
+        terminal_states = {
+            NodeState.RUNNING,
+            NodeState.ERROR,
+            NodeState.TERMINATED,
+            NodeState.STOPPED,
+            NodeState.SUSPENDED,
+            NodeState.PAUSED,
+        }
+
+        try:
+            poll_until(
+                predicate=lambda: self._node_in_state(node, terminal_states),
+                # Overall retry sleep time (not quite the same as timeout since
+                # it doesn't count time spent in the predicate) is just copied
+                # from the default libcloud timeout for wait_until_running.
+                # Maybe some other value would be better.
+                steps=iter([15] * (600 / 15)),
```

Comment:
No action required.

This suggests two improvements to our shared retry logic:

 1. A 'steps' generator that is aware of how much time has passed in the poll and reduces wait time  accordingly
 2. Changing the interface such that the steps generator yields timedeltas